walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,206 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Comparator;
+import java.util.Arrays;
+import java.util.ArrayList;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
-}
+    private static ArrayList<Event> optionalEvents = new ArrayList<>();
+
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        Collection<TimeRange> times = new ArrayList<>();
+
+        if ((int) request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+            return times;
+        } else if (events.size() == 0) {
+            times.add(TimeRange.WHOLE_DAY);
+
+            return times;
+        } else if (events.size() == 1) {
+            boolean included = false;
+
+            for (Event event : events) {
+                if (attends(event, request)) {
+                    included = true;
+                }
+            }
+
+            if (included) {
+                Event[] e = events.toArray(new Event[0]);
+                TimeRange time = e[0].getWhen();
+
+                if (time.start() - TimeRange.START_OF_DAY >= (int) request.getDuration()) {
+                    times.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, time.start(), false));
+                }
+
+                if (TimeRange.END_OF_DAY - time.end() >= (int) request.getDuration()) {
+                    times.add(TimeRange.fromStartEnd(time.end(), TimeRange.END_OF_DAY, true));
+                }
+            } else {
+                times.add(TimeRange.WHOLE_DAY);
+            }
+
+            return times;
+        } else {
+            Event[] e = events.toArray(new Event[0]);
+            Event[] sortedEvents = insertionSort(e, TimeRange.ORDER_BY_START, request);;
```

Comment:
You should be able to use Java's built-in sort rather than having to write your own.

Reply:
I did not know that that existed, thank you for letting me know.